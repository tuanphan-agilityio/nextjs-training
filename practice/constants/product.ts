const SIZES = ['S', 'M', 'L', 'XL'];

export { SIZES };
