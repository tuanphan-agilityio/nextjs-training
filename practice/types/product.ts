interface Product {
  id: number;
  price: number;
  name: string;
  thumbnail: string;
  status: string;
  imgHrefs: string[];
  description: string;
  stock: number;
}

export type { Product };
