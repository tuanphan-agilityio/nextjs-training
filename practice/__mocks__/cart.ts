import { CartProductItem, CartStorage } from '@/types/cart';

const CART_STORAGES: CartStorage[] = [
  { productId: 1, quantity: 2 },
  { productId: 2, quantity: 2 },
  { productId: 3, quantity: 2 },
];

const CART_ITEMS: CartProductItem[] = CART_STORAGES.map(
  ({ productId, quantity }, index) => ({
    productId,
    quantity,
    name: `Product ${index}`,
    thumbnail: `Thumbnail ${index}`,
    price: 30,
    stock: 1305,
  }),
);

export { CART_STORAGES, CART_ITEMS };
