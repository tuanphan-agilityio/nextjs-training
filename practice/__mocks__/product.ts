import { Product } from '@/types/product';

const MOCK_PRODUCT: Product = {
  id: 1,
  name: 'Smart T-shirt',
  price: 120,
  status: 'Best quality',
  thumbnail: '/images/product-2.jpg',
  imgHrefs: ['/images/product-1.jpg', '/images/product-3.jpg'],
  description:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. In eget gravida leo, nec iaculis diam. Nam bibendum mi sed sem finibus ullamcorper.',
  stock: 1305,
};

const MOCK_PRODUCTS: Product[] = Array(3)
  .fill(MOCK_PRODUCT)
  .map((product, index) => ({ ...product, id: index + 1 }));

const MOCK_DEFAULT_PARAMS = {
  _page: 1,
  _limit: 3,
};

export { MOCK_PRODUCT, MOCK_PRODUCTS, MOCK_DEFAULT_PARAMS };
