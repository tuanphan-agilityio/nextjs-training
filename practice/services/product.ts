import { Product } from '@/types/product';
import axiosApp from './axiosApp';
import { ENDPOINTS } from '@/constants/endpoints';
import { Pagination } from '@/types/common';

const getProducts = async (params?: Pagination): Promise<Product[]> => {
  const response: Product[] = await axiosApp.get(ENDPOINTS.PRODUCTS, {
    params,
  });

  return response;
};

const getProduct = async (id: string): Promise<Product> => {
  const response: Product = await axiosApp.get(`${ENDPOINTS.PRODUCTS}/${id}`);

  return response;
};

export { getProducts, getProduct };
