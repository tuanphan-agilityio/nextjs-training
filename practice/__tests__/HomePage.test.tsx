import { render } from '@testing-library/react';

import Home, { HomeProps, getStaticProps } from '@/pages/index';

import { MOCK_DEFAULT_PARAMS, MOCK_PRODUCTS } from '@/__mocks__/product';

jest.mock('@/services/product', () => ({
  getProducts: jest.fn(() => {
    return MOCK_PRODUCTS;
  }),
}));

describe('Home Page', () => {
  const props: HomeProps = {
    params: MOCK_DEFAULT_PARAMS,
    products: MOCK_PRODUCTS,
  };
  it('Home Page matches DOM Snapshot', () => {
    const { container } = render(<Home {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('getStaticProps should return props ', async () => {
    const result = await getStaticProps({});

    expect(result).toStrictEqual({
      props: {
        params: MOCK_DEFAULT_PARAMS,
        products: MOCK_PRODUCTS,
      },
    });
  });
});
