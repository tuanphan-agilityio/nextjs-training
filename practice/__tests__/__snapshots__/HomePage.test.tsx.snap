// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Home Page matches DOM Snapshot 1`] = `
<div>
  <main>
    <section
      class="pt-10 pb-20 bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary md:pt-8 md:pb-14 sm:pt-4 sm:pb-4"
    >
      <div
        class="container"
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            class="flex w-fit text-sm text-primary"
          >
            <li
              class="font-primary-bold"
            >
              Home
            </li>
          </ol>
        </nav>
        <h1
          class="mt-10 text-3xl font-secondary-bold md:mt-6 sm:mt-4"
        >
          Cart
        </h1>
        
      </div>
    </section>
    <section
      class="container py-20 xl:py-16 md:py-12 sm:py-4"
    >
      <h2
        class="mb-6 text-3xl font-secondary-bold"
      >
        T - Shirt
      </h2>
      <div
        class="flex gap-16 xl:flex-col"
      >
        <div>
          <div
            class="flex items-center text-xl"
          >
            <p
              class="font-primary-regular leading-4"
            >
              Filter
            </p>
            <svg
              class="ml-2"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                stroke="#0D1222"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-miterlimit="10"
                stroke-width="1.2"
              >
                <path
                  d="M15.38 7.4a1.9 1.9 0 1 0 0-3.8 1.9 1.9 0 0 0 0 3.8ZM17.27 5.5h3.17M2.6 5.5h10.88M7.9 13.76a1.9 1.9 0 1 0 0-3.8 1.9 1.9 0 0 0 0 3.8ZM9.8 11.86h10.64M2.6 11.86h3.41M15.38 20.12a1.9 1.9 0 1 0 0-3.8 1.9 1.9 0 0 0 0 3.8ZM17.27 18.22h3.17M2.6 18.22h10.88"
                />
              </g>
            </svg>
          </div>
          <div
            class="mt-10 text-md"
          >
            <p
              class="mb-3"
            >
              Categories
            </p>
            <ul
              class="flex gap-6 pb-6 border-b-2 border-solid border-tertiary"
            >
              <li
                class="px-2 py-1 bg-quaternary font-primary-bold text-primary border"
              >
                Men
              </li>
              <li
                class="px-2 py-1 text-secondary border border-tertiary hover:cursor-pointer"
              >
                Women
              </li>
              <li
                class="px-2 py-1 text-secondary border border-tertiary hover:cursor-pointer"
              >
                Children
              </li>
            </ul>
          </div>
        </div>
        <div
          class="text-center"
        >
          <article
            class="flex flex-wrap gap-10 pb-10 sm:gap-4"
          >
            <a
              class="block w-product-card sm:w-full"
              href="/1"
            >
              <section
                class="bg-primary text-secondary shadow text-sm hover:shadow-xl hover:cursor-pointer"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      role="presentation"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjY1IiBoZWlnaHQ9IjI0OCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <noscript />
                  <img
                    alt="Smart T-shirt"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                </div>
                <div
                  class="px-4 pt-4 pb-6 "
                >
                  <div
                    class="flex justify-between items-center mb-2 font-primary-bold xl:flex-wrap"
                  >
                    <h3>
                      Smart T-shirt
                    </h3>
                    <span>
                      120
                      $
                    </span>
                  </div>
                  <p
                    class="mb-3 text-tertiary text-start"
                  >
                    Best quality
                  </p>
                  <div
                    class="flex items-center gap-2 flex-wrap"
                  >
                    <svg
                      fill="none"
                      height="18"
                      width="106"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g
                        fill="#FFD43C"
                      >
                        <path
                          d="M8.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561L1.697 7.127a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM52.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM30.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM74.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM96.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.501.501 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41Z"
                        />
                      </g>
                    </svg>
                    <span
                      class="text-tertiary font-tertiary-regular"
                    >
                      (1k)
                    </span>
                  </div>
                </div>
              </section>
            </a>
            <a
              class="block w-product-card sm:w-full"
              href="/2"
            >
              <section
                class="bg-primary text-secondary shadow text-sm hover:shadow-xl hover:cursor-pointer"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      role="presentation"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjY1IiBoZWlnaHQ9IjI0OCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <noscript />
                  <img
                    alt="Smart T-shirt"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                </div>
                <div
                  class="px-4 pt-4 pb-6 "
                >
                  <div
                    class="flex justify-between items-center mb-2 font-primary-bold xl:flex-wrap"
                  >
                    <h3>
                      Smart T-shirt
                    </h3>
                    <span>
                      120
                      $
                    </span>
                  </div>
                  <p
                    class="mb-3 text-tertiary text-start"
                  >
                    Best quality
                  </p>
                  <div
                    class="flex items-center gap-2 flex-wrap"
                  >
                    <svg
                      fill="none"
                      height="18"
                      width="106"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g
                        fill="#FFD43C"
                      >
                        <path
                          d="M8.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561L1.697 7.127a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM52.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM30.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM74.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM96.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.501.501 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41Z"
                        />
                      </g>
                    </svg>
                    <span
                      class="text-tertiary font-tertiary-regular"
                    >
                      (1k)
                    </span>
                  </div>
                </div>
              </section>
            </a>
            <a
              class="block w-product-card sm:w-full"
              href="/3"
            >
              <section
                class="bg-primary text-secondary shadow text-sm hover:shadow-xl hover:cursor-pointer"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      role="presentation"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjY1IiBoZWlnaHQ9IjI0OCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <noscript />
                  <img
                    alt="Smart T-shirt"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                </div>
                <div
                  class="px-4 pt-4 pb-6 "
                >
                  <div
                    class="flex justify-between items-center mb-2 font-primary-bold xl:flex-wrap"
                  >
                    <h3>
                      Smart T-shirt
                    </h3>
                    <span>
                      120
                      $
                    </span>
                  </div>
                  <p
                    class="mb-3 text-tertiary text-start"
                  >
                    Best quality
                  </p>
                  <div
                    class="flex items-center gap-2 flex-wrap"
                  >
                    <svg
                      fill="none"
                      height="18"
                      width="106"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g
                        fill="#FFD43C"
                      >
                        <path
                          d="M8.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561L1.697 7.127a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM52.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM30.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM74.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM96.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.501.501 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41Z"
                        />
                      </g>
                    </svg>
                    <span
                      class="text-tertiary font-tertiary-regular"
                    >
                      (1k)
                    </span>
                  </div>
                </div>
              </section>
            </a>
          </article>
          <button
            class="inline-flex justify-center items-center w-[180px] h-12 bg-primary font-primary-regular text-md hover:opacity-80 bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary text-center"
          >
            Load More
          </button>
        </div>
      </div>
    </section>
  </main>
</div>
`;
